### Exercícios de lógica de programação envolvendo estruturas de repetição.

**1. Escreva um programa que imprime todos os numeros de 0 até 50, incluindo-os.**

In [1]:
i=0
while(i<=50):
    print(i, end=' ')
    i=i+1

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 

In [2]:
for i in range(51):
    print(i, end=' ')

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 

**2. Modifique o programa anterior de forma que este imprima apenas os números que são pares.**

In [3]:
i=0
while(i<=50):
    print(i, end=' ')
    i=i+2

0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 

In [4]:
i=0
while(i<=50):
    if i%2==0:
        print(i, end=' ')
    i=i+1

0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 

In [5]:
for i in range(0,51,2):
    print(i, end=' ')

0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 

In [6]:
for i in range(0,51):
    if i%2 ==0:
        print(i, end=' ')

0 2 4 6 8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50 

**3. Escreva um programa para contar a quantidade de números pares entre dois números quaisquer fornecidos pelo usuário.**

In [1]:
# solução incluindo os limites usando while
number1 = int(input('Digite um número inteiro: '))
number2 = int(input('Digite um outro número inteiro: '))
even_counter = 0

if number1 < number2:
    i = number1
    while i <= number2:
        if i%2 == 0:
            even_counter += 1
        i += 1
else:
    i = number2
    while i <= number1:
        if i%2 == 0:
            even_counter += 1
        i += 1
print(f'Existem {even_counter} números pares entre {number1} e {number2}.')

Existem 4 números pares entre 2 e 8.


In [3]:
# solução excluindo os limites usando while
number1 = int(input('Digite um número inteiro: '))
number2 = int(input('Digite um outro número inteiro: '))
even_counter = 0

if number1 < number2:
    i = number1+1
    while i < number2:
        if i%2 == 0:
            even_counter += 1
        i += 1
else:
    i = number2+1
    while i < number1:
        if i%2 == 0:
            even_counter += 1
        i += 1
print(f'Existem {even_counter} números pares entre {number1} e {number2}.')

Existem 2 números pares entre 8 e 2.


In [2]:
# solução incluindo os limites usando for
number1 = int(input('Digite um número inteiro: '))
number2 = int(input('Digite um outro número inteiro: '))
even_counter = 0

if number1 < number2:
    for i in range (number1, number2+1):
        if i%2 == 0:
            even_counter += 1
else:
    for i in range (number2, number1+1):
        if i%2 == 0:
            even_counter += 1
print(f'Existem {even_counter} números pares entre {number1} e {number2}.')

Existem 4 números pares entre 8 e 2.


In [5]:
# solução excluindo os limites usando for
number1 = int(input('Digite um número inteiro: '))
number2 = int(input('Digite um outro número inteiro: '))
even_counter = 0

if number1 < number2:
    for i in range (number1+1, number2):
        if i%2 == 0:
            even_counter += 1
else:
    for i in range (number2+1, number1):
        if i%2 == 0:
            even_counter += 1
print(f'Existem {even_counter} números pares entre {number1} e {number2}.')

Existem 2 números pares entre 2 e 8.


**4. Escreva um programa para calcular o fatorial de um número fornecido pelo usuário.**

In [7]:
number = int(input('Digite um número inteiro positivo: '))
factorial = 1

for i in range(1, number+1):
    factorial *= i
print(f'O fatorial de {number} é {factorial}.')

O fatorial de 5 é 120.


In [4]:
number = int(input('Digite um número inteiro positivo: '))
factorial = 1

for i in range(number, 0, -1):
    factorial *= i
print(f'O fatorial de {number} é {factorial}.')

O fatorial de 5 é 120.


In [5]:
number = int(input('Digite um número inteiro positivo: '))
factorial = 1
i = number
while i>= 1:
    factorial *= i
    i -= 1
print(f'O fatorial de {number} é {factorial}.')

O fatorial de 5 é 120.


**5. Faça um programa que peça dois números, base e expoente, calcule e mostre o primeiro número elevado ao segundo número.** 
- Não utilize a função de potência da linguagem.

In [1]:
number1 = int(input('Digite o valor da base '))
number2 = int(input('Digite o valor do expoente '))
result = 1

if number2 >= 0:
    for i in range (number2):
        result *= number1
else:
    for i in range (-(number2)):
        result *= 1/number1
    
print(f'O número {number1} elevado a {number2} é {result}.')

O número 2 elevado a -5 é 0.03125.


In [2]:
number1 = int(input('Digite o valor da base '))
number2 = int(input('Digite o valor do expoente '))
result = 1

i = 1
if number2 >= 0:
    while i <= number2:
        result *= number1
        i += 1
else:
    while i <= -number2:
        result *= 1/number1
        i += 1
    
print(f'O número {number1} elevado a {number2} é {result}.')

O número 2 elevado a 0 é 1.


**6. Desenvolva um gerador de tabuada, capaz de gerar a tabuada de qualquer número inteiro entre 1 a 10. O usuário deve informar de qual numero ele deseja ver a tabuada. A saída deve ser conforme o exemplo abaixo:**

<div align="center">
    <img src=./images/004_6.png width="10%">
</div>

In [8]:
number = int(input('Digite um número inteiro: '))
print(f'Tabuada do número {number}: ')
for i in range(1,11):
    print(f'{number} X {i} = {number*i}')

Tabuada do número 5: 
5 X 1 = 5
5 X 2 = 10
5 X 3 = 15
5 X 4 = 20
5 X 5 = 25
5 X 6 = 30
5 X 7 = 35
5 X 8 = 40
5 X 9 = 45
5 X 10 = 50


**7. A prefeitura de uma cidade deseja fazer uma pesquisa entre seus habitantes. Faça um algoritmo para coletar e armazenar dados sobre o salário e número de filhos de cada habitante e após as leituras, escrever:**

**a)** Média de salário da população

**b)** Média do número de filhos

**c)** Maior salário dos habitantes

**d)** Percentual de pessoas com salário menor que R$ 150,00

**Obs.:** O final da leituras dos dados se dará com a entrada de um “salário negativo”.

In [10]:
population_count = 0
salary_count = 0.0
children_count = 0
biggest_salary = 0
less_150_count = 0

while True:
  salary = float(input(f'Digite o salário do habitante {population_count+1}: ')) 
  if salary < 0: break
  if salary > biggest_salary:
    biggest_salary = salary
  if salary < 150.0:
    less_150_count += 1
  salary_count += salary
  filhos = float(input(f'Digite o número de filhos do habitante {population_count+1}: '))
  children_count +=filhos
  population_count +=1
if population_count>0:
  print(f'O salário médio da população é de: R$ {salary_count/population_count:.2f}.')  
  print(f'O número médio de filhos é: {children_count/population_count:.2f}.')  
  print(f'O maior salário é: R$ {biggest_salary}.')  
  print(f'O percentual de pessoas com um salário inferior a 150 é: {100*less_150_count/population_count:.2f} %.')

O salário médio da população é de: R$ 5040.00.
O número médio de filhos é: 2.33.
O maior salário é: R$ 10000.0.
O percentual de pessoas com um salário inferior a 150 é: 33.33 %.


**8. Escreva um programa que leia um número inteiro e calcule a soma de todos os divisores desse número, com exceção dele próprio.** 

Ex: a soma dos divisores do número 66 é 1 + 2 + 3 + 6 + 11 + 22 + 33 = 78.

In [8]:
number = int(input('Insira um número inteiro para calcular a soma de seus divisores: '))
accumulator = 0
for i in range(1, number):
    if number%i == 0:
        accumulator += i
print(f'A soma dos divisores do número {number} é: {accumulator}.')

A soma dos divisores do número 66 é: 78.


**9. Em Matemática, o número harmônico designado por sendo a soma da série harmônica demonstrada na imagem abaixo. Faça um programa que leia um valor n inteiro e positivo e apresente o valor de *H(n)*.**

<div align="center">
    <img src=./images/004_9.png width="25%">
</div>

In [10]:
number = int(input('Digite um número inteiro para calcular o número harmônico: '))
accumulator = 0
for i in range(1, number+1):
    accumulator += 1/i
print(f'O valor de H({number}) é: {accumulator:.2f}.')

O valor de H(5) é: 2.28.


**10. A série de Fibonacci é formada pela seqüência 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, ... Faça um programa capaz de gerar a série até o enésimo termo.**

In [5]:
number = int(input('Digite o enésimo termo para calcular a sequência de Fibonacci: '))
print(f'Fibonacci até o {number}º termo: ', end='') 
if (number == 1):
    print("1")
elif(number==2):
    print("1, 1")
else:
    print("1, 1", end='')
    prev_prev = 1
    prev = 1
    for i in range(3, number+1):
        current_number = prev_prev + prev
        print(f', {current_number}', end='')
        prev_prev = prev
        prev = current_number
    print()
  

Fibonacci até o 10º termo: 1, 1, 2, 3, 5, 8, 13, 21, 34, 55


**11. A série de Fibonacci é formada pela seqüência 1, 1, 2, 3, 5, 8, 13, 21, 34, 55,... Faça um programa que gere a série enquanto o valor seja maior que 500.**

In [4]:
print("1, 1", end="")
prev_prev = 1
prev = 1
while True:
    current_number = prev_prev + prev
    if current_number > 500: 
        break
    print(',' ,current_number, end="")
    prev_prev = prev
    prev = current_number

1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377

**12. Faça um programa que calcule o fatorial de um número inteiro fornecido pelo usuário.**

Ex.: 5!=5.4.3.2.1=120. 

A saída deve exatamente como o exemplo abaixo:

<div align="center">
    <img src=./images/004_12.png width="25%">
</div>

In [16]:
number = int(input('Digite um número inteiro positivo: '))
factorial = 1

print(f'Fatorial de: {number} \n{number}! = ', end="")

for i in range(number, 1, -1):
    print(i, end=" . ")
    factorial *= i
print(f'1 = {factorial}')

Fatorial de: 5 
5! = 5 . 4 . 3 . 2 . 1 = 120
